### PR TITLE
CLM-22527: Add documentation to explain plugin architecture and setup

### PR DESCRIPTION
**Pull request by DanielRuaMadrid**
Add more documentation to explain plugin architecture and how to set up the plugin properly

### CODE REVIEW

File: docs/overview.md [hunk 1]
```diff
@@ -53,6 +53,51 @@ JFrog Artifactory 'properties' are used to store component metadata.
 3) ALLOW / DENY status received and saved to cache
 4) Access to component is allowed or denied
 
+## How do the Firewall for JFrog Artifactory works
+
+Artifactory's plugins allow extending Artifactory's behaviour to perform [tasks](https://www.jfrog.com/confluence/display/JFROG/User+Plugins#UserPlugins-AboutPlugins) 
+as handling and manipulating download events, perform searches, query artifacts and metadata, etc. The plugins need to be
+written in Groovy but is possible having an entrypoint file in Groovy a from there call Java classes. In this plugin the
```
Comment (fernau) [inline docs/overview.md:60]:
```suggestion
written in Groovy but is possible having an entrypoint file in Groovy and from there call Java classes. In this plugin the 
```

**Comment by fernau** [comment]
You could add a section on where to add unit tests and how to run them

**Comment by fernau** [comment]
maybe we could add something about how to debug this. it can only be debugged when running a unit test, since we don't have an bundled artifactory server. maybe just say that you can place a breakpoint in your debugger to stop the execution and step through the code.

**Comment by fernau** [comment]
Hi, one more thing, coud you add some build instruction. e.g. version of java supported, build tool requirements. Thanks

File: docs/overview.md [hunk 2]
```diff
@@ -53,6 +53,51 @@ JFrog Artifactory 'properties' are used to store component metadata.
 3) ALLOW / DENY status received and saved to cache
 4) Access to component is allowed or denied
 
+## How do the Firewall for JFrog Artifactory works
```
Comment (vladt) [inline docs/overview.md:56]: How do the Firewall for JFrog Artifactory works -> How does the Firewall for JFrog Artifactory work

File: docs/overview.md [hunk 3]
```diff
@@ -53,6 +53,51 @@ JFrog Artifactory 'properties' are used to store component metadata.
...
+- **Download handler event types**
+  - beforeDownload: Makes a call to Firewall to process the component and determine if it is possible to download the component.
+- **Storage handler event types**
+  - beforePropertyCreate: Checks property access.
+  - beforePropertyDelete: Checks property access.
+  - afterDelete: Calls Firewall to remove component from internal database.
+
+An example of the workflow of the plugin is the next:
+
+![Sequence Diagram - Download od components](sequence_diagram.png "Sequence diagram - download components")
```
Comment (vladt) [inline docs/overview.md:78]: In the diagram, "Returns 5xx" -> "Returns 403"

File: docs/overview.md [hunk 4]
```diff
@@ -53,6 +53,51 @@ JFrog Artifactory 'properties' are used to store component metadata.
...
+
+Having installed and configured correctly the plugin the only next thing required is restart Artifactory. This could vary depending on the installation method.
+In Unix systems and configured to run Artifactory as a service we will only to run the following commands:
+
+`systemctl stop artifactory.service`
+`systemctl start artifactory.service`
+
 ## About JFrog Artifactory plugins
 
 To keep this plugin as stable as possible across JFrog Artifactory releases we only use functionality provided by the
```
Comment (vladt) [inline docs/overview.md:104]: The "Build dependencies on external projects" section needs to be removed. Those repos don't exist anymore.

Comment (vladt) [reply docs/overview.md:104]: The "Project directory layout" section needs to be updated to match the new layout.

File: docs/overview.md (hunk 3 above)
Comment (vladt) [inline docs/overview.md:78]: In the diagram, I would replace "FWFA" with the full name.

File: docs/overview.md [hunk 5]
```diff
@@ -69,3 +115,35 @@ The plugin requires dependencies from the [jcenter repository](https://jcenter.b
...
+
+## Build instructions
+
+For building this plugin is required to have installed the following tools:
+
+- Maven 2 or higher
+- Java - JDK 11
+- [Groovy](https://groovy-lang.org/)
+
+For building and packaging the plugin is required to execute the `package` maven goal using the command-line or through an IDE.
```
Comment (vladt) [inline docs/overview.md:127]: To build and package the plugin, use the `package` maven goal:

File: docs/overview.md [hunk 6]
```diff
@@ -69,3 +115,35 @@ The plugin requires dependencies from the [jcenter repository](https://jcenter.b
...
+## How to run the tests
+
+Running tests is possible in different ways using the maven goal from the command-line or using an IDE that uses the maven goal underneath.
+Using the command line is as simple as running `mvn test` command in the `firewall` module. For running the tests from an IDE please refer to the documentation of you favorite IDE
+
+- [IntelliJ IDEA](https://www.jetbrains.com/help/idea/performing-tests.html#run-tests-before-commit)
+- [Netbeans](https://netbeans.apache.org/tutorials/nbm-test.html)
+- [Eclipse](https://wiki.eclipse.org/Eclipse/Testing)
+
+## How to debug
```
Comment (vladt) [inline docs/overview.md:146]: I would remove this section. The plugin can be debugged like any other java application.

**Comment by thingstiger** [comment]
Hi, There is one more thing we need to add. External users will be unable to build this without the PAPI.jar file which you cannot get from central.
We need to add instructions on how to get that file an how to add it to your local cache or repo server so you can build the project

File: docs/overview.md (hunk 2 above)
Comment (DanielRuaMadrid) [reply docs/overview.md]: Fixed https://github.com/sonatype/firewall-for-jfrog-artifactory/pull/9/commits/c98363e46afa7b3c1c8c2bbac5ff874530b98257

File: docs/overview.md (hunk 3 above)
Comment (DanielRuaMadrid) [reply docs/overview.md:78]: Fixed https://github.com/sonatype/firewall-for-jfrog-artifactory/pull/9/commits/c98363e46afa7b3c1c8c2bbac5ff874530b98257

File: docs/overview.md (hunk 5 above)
Comment (DanielRuaMadrid) [reply docs/overview.md]: Fixed https://github.com/sonatype/firewall-for-jfrog-artifactory/pull/9/commits/c98363e46afa7b3c1c8c2bbac5ff874530b98257

File: docs/overview.md (hunk 6 above)
Comment (DanielRuaMadrid) [reply docs/overview.md:146]: I changed this section a little bit to make it clearer and include the two ways to debug the plugin https://github.com/sonatype/firewall-for-jfrog-artifactory/pull/9/commits/4ce2afabd2b1037a5d684a503ed3067cd88d7d66

File: docs/overview.md [hunk 7]
```diff
@@ -53,19 +53,99 @@ JFrog Artifactory 'properties' are used to store component metadata.
...
+- lib/nexus-iq-artifactory-plugin.jar: All the code called by the plugin.
+
+To install the plugin you need to extract the zip file into JFrog Artifactory's `/plugins` directory. The location of this folder
+could change depending on which Artifactory version you are using. [For more details see](https://help.sonatype.com/fw/getting-started/jfrog-artifactory-setup).
+
+When you have the files extracted into the `/plugins` directory the `firewall.properties.example` file needs to be renamed to `firewall properties` and
+configured to point to the IQ server. [See configuration example](https://help.sonatype.com/fw/getting-started/jfrog-artifactory-setup#JFrogArtifactorySetup-configurationConfiguration).
+
+Having installed and configured correctly the plugin the only next thing required is restart Artifactory. This could vary depending on the installation method.
+In Unix systems and configured to run Artifactory as a service we will only to run the following commands:
```
Comment (vladt) [inline docs/overview.md:96]: we will only to -> we will only need to

File: docs/overview.md [hunk 8]
```diff
@@ -53,19 +53,99 @@ JFrog Artifactory 'properties' are used to store component metadata.
...
+
+`systemctl stop artifactory.service`
```
Comment (vladt) [inline docs/overview.md:98]: Simpler:
systemctl restart artifactory.service

File: docs/overview.md (hunk 4 above)
Comment (vladt) [reply docs/overview.md:104]: I think the "Project directory layout" section still needs to be updated to match the new layout.

File: docs/overview.md [hunk 9]
```diff
@@ -53,19 +53,99 @@ JFrog Artifactory 'properties' are used to store component metadata.
...
 * **plugin**: Contains the groovy file implementing the plugin
-* **vars**: build scripts
+
+## Build instructions
+
+For building this plugin is required to have installed the following tools:
+
+- Maven 2 or higher
+- Java - JDK 11
+- [Groovy](https://groovy-lang.org/)
```
Comment (vladt) [inline docs/overview.md:120]: I don't believe you need groovy installed.

File: docs/overview.md [hunk 10]
```diff
@@ -53,19 +53,99 @@ JFrog Artifactory 'properties' are used to store component metadata.
...
+- [IntelliJ IDEA](https://www.jetbrains.com/help/idea/performing-tests.html#run-tests-before-commit)
+- [Netbeans](https://netbeans.apache.org/tutorials/nbm-test.html)
+- [Eclipse](https://wiki.eclipse.org/Eclipse/Testing)
+
+## How to debug
+
+To debug the plugin is possible to do it in two different ways:
+
+- Remote debug the plugin installed in the Artifactory Server Pro
+- Set up your favorite IDE (see prior section) to run tests and set breakpoints to stop the execution and ste through the code.
```
Comment (vladt) [inline docs/overview.md:151]: ste->step

Comment (DanielRuaMadrid) [reply docs/overview.md]: Fixed https://github.com/sonatype/firewall-for-jfrog-artifactory/pull/9/commits/7df067a5749396d695d5fbe07422bf72b76527d6

File: docs/overview.md (hunk 9 above)
Comment (DanielRuaMadrid) [reply docs/overview.md]: Fixed https://github.com/sonatype/firewall-for-jfrog-artifactory/pull/9/commits/7df067a5749396d695d5fbe07422bf72b76527d6

File: docs/overview.md (hunk 8 above)
Comment (DanielRuaMadrid) [reply docs/overview.md]: Fixed https://github.com/sonatype/firewall-for-jfrog-artifactory/pull/9/commits/7df067a5749396d695d5fbe07422bf72b76527d6

File: docs/overview.md (hunk 7 above)
Comment (DanielRuaMadrid) [reply docs/overview.md]: Fixed https://github.com/sonatype/firewall-for-jfrog-artifactory/pull/9/commits/7df067a5749396d695d5fbe07422bf72b76527d6

File: docs/overview.md (hunk 4 above)
Comment (DanielRuaMadrid) [reply docs/overview.md:102]: Actually it matches with the directory layout. I just add more information https://github.com/sonatype/firewall-for-jfrog-artifactory/pull/9/commits/7df067a5749396d695d5fbe07422bf72b76527d6

Comment (vladt) [reply docs/overview.md:102]: > Actually it matches with the directory layout.

You're right. Sorry.

File: docs/overview.md [hunk 11]
```diff
@@ -53,19 +53,96 @@ JFrog Artifactory 'properties' are used to store component metadata.
 3) ALLOW / DENY status received and saved to cache
```
Comment (vladt) [inline docs/overview.md:53]: In this file, "then an CancelException" -> "then a CancelException"

File: docs/overview.md [hunk 12]
```diff
@@ -53,19 +53,96 @@ JFrog Artifactory 'properties' are used to store component metadata.
...
+- nexusFirewallForArtifactoryPlugin.groovy: the plugin itself. This is the file called by Artifactory.
+- lib/nexus-iq-artifactory-plugin.jar: All the code called by the plugin.
+
+To install the plugin you need to extract the zip file into JFrog Artifactory's `/plugins` directory. The location of this folder
+could change depending on which Artifactory version you are using. [For more details see](https://help.sonatype.com/fw/getting-started/jfrog-artifactory-setup).
+
+When you have the files extracted into the `/plugins` directory the `firewall.properties.example` file needs to be renamed to `firewall properties` and
+configured to point to the IQ server. [See configuration example](https://help.sonatype.com/fw/getting-started/jfrog-artifactory-setup#JFrogArtifactorySetup-configurationConfiguration).
+
+Having installed and configured correctly the plugin the only next thing required is restart Artifactory. This could vary depending on the installation method.
```
Comment (vladt) [inline docs/overview.md:95]: is restart Artifactory -> is to restart Artifactory

File: docs/overview.md [hunk 13]
```diff
@@ -53,19 +53,96 @@ JFrog Artifactory 'properties' are used to store component metadata.
...
+`nexus-iq-artifactory-plugin-$PLUGIN_VERSION-.zip`. When you unzip this file you encounter the following structure:
+
+- firewall.properties.example: All the properties related to the firewall connection and repository configuration.
+- nexusFirewallForArtifactoryPlugin.groovy: the plugin itself. This is the file called by Artifactory.
+- lib/nexus-iq-artifactory-plugin.jar: All the code called by the plugin.
+
+To install the plugin you need to extract the zip file into JFrog Artifactory's `/plugins` directory. The location of this folder
+could change depending on which Artifactory version you are using. [For more details see](https://help.sonatype.com/fw/getting-started/jfrog-artifactory-setup).
+
+When you have the files extracted into the `/plugins` directory the `firewall.properties.example` file needs to be renamed to `firewall properties` and
```
Comment (vladt) [inline docs/overview.md:92]: directory the -> directory, the

Comment (vladt) [reply docs/overview.md:92]: firewall properties -> firewall.properties

File: docs/overview.md [hunk 14]
```diff
@@ -53,19 +53,96 @@ JFrog Artifactory 'properties' are used to store component metadata.
...
+configured to point to the IQ server. [See configuration example](https://help.sonatype.com/fw/getting-started/jfrog-artifactory-setup#JFrogArtifactorySetup-configurationConfiguration).
+
+Having installed and configured correctly the plugin the only next thing required is restart Artifactory. This could vary depending on the installation method.
+In Unix systems and configured to run Artifactory as a service we will only need to run the following commands:
```
Comment (vladt) [inline docs/overview.md:96]: remove "and"

File: docs/overview.md [hunk 15]
```diff
@@ -53,19 +53,96 @@ JFrog Artifactory 'properties' are used to store component metadata.
...
-* **vars**: build scripts
+* **docs**: This directory. Contains the technical information related to the plugin.
+* **firewall**: Builds the firewall-plugin's jar file. All code related to the communication with Nexus IQ.
+* **plugin**: Contains the groovy file implementing the plugin. It follows the [plugin template](https://www.jfrog.com/confluence/display/JFROG/User+Plugins#UserPlugins-PluginTemplateSource).
+
+## Build instructions
+
+For building this plugin is required to have installed the following tools:
+
+- Maven 2 or higher
```
Comment (vladt) [inline docs/overview.md:116]: I tried Maven 2 and it did not work. Let's say "Maven 3.6.x or higher"

File: docs/overview.md (hunk 11 above)
Comment (DanielRuaMadrid) [reply docs/overview.md:53]: Fixed https://github.com/sonatype/firewall-for-jfrog-artifactory/pull/9/commits/7fc2c95b43ed32d2ac451f7af8bc427e3a91205e

File: docs/overview.md (hunk 12 above)
Comment (DanielRuaMadrid) [reply docs/overview.md]: Fixed https://github.com/sonatype/firewall-for-jfrog-artifactory/pull/9/commits/7fc2c95b43ed32d2ac451f7af8bc427e3a91205e

File: docs/overview.md (hunk 14 above)
Comment (DanielRuaMadrid) [reply docs/overview.md]: Fixed https://github.com/sonatype/firewall-for-jfrog-artifactory/pull/9/commits/7fc2c95b43ed32d2ac451f7af8bc427e3a91205e

File: docs/overview.md (hunk 15 above)
Comment (DanielRuaMadrid) [reply docs/overview.md]: Fixed https://github.com/sonatype/firewall-for-jfrog-artifactory/pull/9/commits/7fc2c95b43ed32d2ac451f7af8bc427e3a91205e

**Comment by vladt** [comment]
> Hi, There is one more thing we need to add. External users will be unable to build this without the PAPI.jar file which you cannot get from central.
We need to add instructions on how to get that file an how to add it to your local cache or repo server so you can build the project. I guess you have to do something like this if you don't have a local repo server:
mvn install:install-file -Dfile=/tmp/artifactory-papi-7.12.5.jar -DgroupId=org.artifactory -DartifactId=artifactory-papi -Dversion=7.12.5 -Dpackaging=jar

Good point. Thank you. Incorporated in the doc.

**Review by laas29 — APPROVED** [review]

**Pull request merged**